README.md

# sparen

Create richer more helpful logs using terminal coloring, file saving,
graph plotting, and ascii drawing.

``` javascript

    const sparen = require('sparen');
    sparen.log('This is a message');

```

---------------------------------------------------------------------
## Table of contents

* [Install](#install)
* [Examples](#examples)
* [References](#references)

&nbsp;

---------------------------------------------------------------------
## Install

    $ npm install sparen

&nbsp;


---------------------------------------------------------------------
## Examples

### Simple logging
``` javascript

    const sparen = require('sparen');
    const Log = sparen.log;

    // Save to a disk file
    // sparen.obj.setLogFile('./sparen.log');

    // Output to stderr instead of stdout
    // sparen.obj.setLogFunction(console.error);

    // Add color filters
    sparen.addLogFilters('info:blue,warning:yellow,error:red');

    Log('This is information');
    Log('This is a warning');
    Log('This is an error');

    let lst = ['this', 'is', 'a', 'list'];
    let arr = {'a': 'b', 'c': 'd', 'e': {'f': ['g', 'h', 'i'], 'j': 'k'}};

    Log('This is a list: ', lst);
    Log('This is an array: ', arr);

    try
    {
        thisis = anerror;
    }
    catch(e)
    {
        Log('CAPTURED ERROR', e);
    }

```
![plot](https://raw.githubusercontent.com/wheresjames/sparen-js/master/img/logtxt.png)

### Array plots
``` javascript

    let plot = sparen.initArray(200, (i)=>Math.sin(i * Math.PI * 6 / 200) * 10);
    Log('PLOT\n' + sparen.plotArray(plot) + "\n");

```
![plot](https://raw.githubusercontent.com/wheresjames/sparen-js/master/img/logplot.png)

### Canvas drawing
``` javascript

    let canv = new sparen.Canvas(80, 25, 2);

    canv.line(2, 2, 40, 2);
    canv.line(77, 2, 77, 12);

    canv.rect(4, 4, 20, 10);
    canv.line(5, 5, 19, 9);

    canv.fillRect(25, 6, 45, 14);

    canv.circle(60, 6, 5);
    canv.arc(60, 6, 5, 0, 180, '*');

    canv.text(60, 6, "2");

    canv.rect(10, 15, 50, 23);
    canv.textBox(11, 15, 49, 23, 'This\tis a lot of text just to see how well it fits in to the'
                                +' specified box, I could go on and on, and I will, because the'
                                +' point here is to make a really long string and not to not'
                                +' freak out people that do not like to see a lot of text.');

    canv.rect(55, 15, 79, 23);
    canv.textBox(55, 15, 79, 23, 'ThisWordIsJustTooLongToFitOnOneLineAndMustBeForcefullySplit.\n'
                                +' So \n be \n it.');

    Log("First Example\n" + canv.toString());

```
![plot](https://raw.githubusercontent.com/wheresjames/sparen-js/master/img/logdraw.png)

### Canvas line drawing
``` javascript

    canv = new sparen.Canvas(80, 25, 1);

    canv.rect(4, 4, 20, 10);

    canv.line(2, 6, 25, 6);
    canv.line(4, 8, 20, 8);

    canv.line(10, 2, 10, 12);
    canv.line(16, 4, 16, 10);

    canv.rect(44, 4, 60, 10);
    canv.line(42, 4, 65, 4);
    canv.line(42, 10, 68, 10);

    canv.rect(24, 14, 40, 20);
    canv.line(24, 12, 24, 22);
    canv.line(40, 12, 40, 24);

    canv.rect(45, 14, 65, 20);
    canv.rect(50, 16, 70, 22);
    canv.rect(55, 18, 75, 24);

    canv.rect(2, 14, 10, 18);
    canv.rect(10, 14, 18, 18);
    canv.rect(2, 18, 10, 22);
    canv.rect(10, 18, 18, 22);

    Log("Second Example\n" + canv.toString());

```
![plot](https://raw.githubusercontent.com/wheresjames/sparen-js/master/img/loglines.png)


&nbsp;


---------------------------------------------------------------------
## References

- Node.js
    - https://nodejs.org/

- npm
    - https://www.npmjs.com/


CHANGES

# v0.2.2 - 2022-01-25

[+] Relative and full path options

test/demo.js
#!/usr/bin/env nodejs
'use strict';

const sparen = require('sparen');

const Log = sparen.log;

sparen.log('This is a message');

sparen.obj.addLogFilters('info:blue,warning:yellow,error:red');
Log('This is information');
Log('This is a warning');
Log('This is an error');

let lst = ['this', 'is', 'a', 'list'];
let arr = {'a': 'b', 'c': 'd', 'e': {'f': ['g', 'h', 'i'], 'j': 'k'}};

Log('This is a list: ', lst);
Log('This is an array: ', arr);

try
{
    thisis = anerror;
}
catch(e)
{
    Log('CAPTURED ERROR', e);
}

//-------------------------------------------------------------------
// PLOT

let plot = sparen.initArray(200, (i)=>Math.sin(i * Math.PI * 6 / 200) * 10);
Log('PLOT\n' + sparen.plotArray(plot) + "\n");

//-------------------------------------------------------------------
// Canvas Example 1

let canv = new sparen.Canvas(80, 25, 2);

canv.line(2, 2, 40, 2);
canv.line(77, 2, 77, 12);

canv.rect(4, 4, 20, 10);
canv.line(5, 5, 19, 9);

canv.fillRect(25, 6, 45, 14);

canv.circle(60, 6, 5);
canv.arc(60, 6, 5, 0, 180, '*');

canv.text(60, 6, "2");

canv.rect(10, 15, 50, 23);
canv.textBox(11, 15, 49, 23, 'This\tis a lot of text just to see how well it fits in to the'
                            +' specified box, I could go on and on, and I will, because the'
                            +' point here is to make a really long string and not to not'
                            +' freak out people that do not like to see a lot of text.');

canv.rect(55, 15, 79, 23);
canv.textBox(55, 15, 79, 23, 'ThisWordIsJustTooLongToFitOnOneLineAndMustBeForcefullySplit.\n'
                            +' So \n be \n it.');

Log("First Example\n" + canv.toString());


//-------------------------------------------------------------------
// Canvas Example 2

canv = new sparen.Canvas(80, 25, 1);

canv.rect(4, 4, 20, 10);

canv.line(2, 6, 25, 6);
canv.line(4, 8, 20, 8);

canv.line(10, 2, 10, 12);
canv.line(16, 4, 16, 10);

canv.rect(44, 4, 60, 10);
canv.line(42, 4, 65, 4);
canv.line(42, 10, 68, 10);

canv.rect(24, 14, 40, 20);
canv.line(24, 12, 24, 22);
canv.line(40, 12, 40, 24);

canv.rect(45, 14, 65, 20);
canv.rect(50, 16, 70, 22);
canv.rect(55, 18, 75, 24);

canv.rect(2, 14, 10, 18);
canv.rect(10, 14, 18, 18);
canv.rect(2, 18, 10, 22);
canv.rect(10, 18, 18, 22);

Log("Second Example\n" + canv.toString());


test/test.js
#!/usr/bin/env nodejs
'use strict';

const assert = require('assert');
const sparen = require('sparen');

const Log = sparen.log;
const Err = sparen.log;
const Fmt = JSON.stringify;

function test_1()
{
    Log('This is a message');

    sparen.obj.addLogFilters('info:blue,warning:yellow,error:red');
    Log('This is information');
    Log('This is a warning');
    Log('This is an error');

    let lst = ['this', 'is', 'a', 'list'];
    let arr = {'a': 'b', 'c': 'd', 'e': {'f': ['g', 'h', 'i'], 'j': 'k'}};

    Log('This is a list: ', lst);
    Log('This is an array: ', arr);

    try
    {
        thisis = anerror;
    }
    catch(e)
    {
        Err('CAPTURED EXCEPTION', e);
    }

    let sz = 200;
    let data = [];
    for (let i = 0; i < sz; i++)
        data.push(Math.sin(i * Math.PI * 6 / sz) * 10);

    function initArray(sz, f)
    {   let a = [];
        for (let i = 0; i < sz; i++)
            a.push(f(i));
        return a;
    }
}

function test_2()
{
    let plot = sparen.initArray(200, (i)=>Math.sin(i * Math.PI * 6 / 200) * 10);
    Log('PLOT\n' + sparen.plotArray(plot));
}

function test_3()
{
    let canv = new sparen.Canvas(80, 25, 2);

    canv.line(2, 2, 40, 2);
    canv.line(77, 2, 77, 12);

    canv.rect(4, 4, 20, 10);
    canv.line(5, 5, 19, 9);

    canv.fillRect(25, 6, 45, 14);

    canv.circle(60, 6, 5);
    canv.arc(60, 6, 5, 0, 180, '*');

    canv.text(60, 6, "2");

    canv.rect(10, 15, 50, 23);
    canv.textBox(11, 15, 49, 23, 'This\tis a lot of text just to see how well it fits into the'
                                +' specified box, I could go on and on, and I will, because the'
                                +' point here is to make a really long string and not to not'
                                +' freak out people that do not like to see a lot of text.');

    canv.rect(55, 15, 79, 23);
    canv.textBox(55, 15, 79, 23, 'ThisWordIsJustTooLongToFitOnOneLineAndMustBeForcefullySplit.\n'
                                +' So \n be \n it.');

    Log("First Example\n" + canv.toString());
}

function test_4()
{
    let canv = new sparen.Canvas(80, 25, 1);

    canv.rect(4, 4, 20, 10);

    canv.line(2, 6, 25, 6);
    canv.line(4, 8, 20, 8);

    canv.line(10, 2, 10, 12);
    canv.line(16, 4, 16, 10);

    canv.rect(44, 4, 60, 10);
    canv.line(42, 4, 65, 4);
    canv.line(42, 10, 68, 10);

    canv.rect(24, 14, 40, 20);
    canv.line(24, 12, 24, 22);
    canv.line(40, 12, 40, 24);

    canv.rect(45, 14, 65, 20);
    canv.rect(50, 16, 70, 22);
    canv.rect(55, 18, 75, 24);

    canv.rect(2, 14, 10, 18);
    canv.rect(10, 14, 18, 18);
    canv.rect(2, 18, 10, 22);
    canv.rect(10, 18, 18, 22);

    Log("Second Example\n" + canv.toString());
}

function test_5()
{
    let mins  = 60;
    let hrs   = 60 * 60;
    let days  = 24 * 60 * 60;
    let years = 365 * 24 * 60 * 60;

    let ts, t = [
            4 * hrs + 12 * mins + 23,                           // 04:12:23
            4 * hrs + 12 * mins + 23 + 0.340,                   // 04:12:23.340
            123 * days + 4 * hrs + 2 * mins + 6,                // 123 days 04:02:06
            4 * years + 123 * days + 4 * hrs + 2 * mins + 6,    // 4 years, 123 days, 04:02:06
        ];

    ts = sparen.formatInterval(t[0]);
    Log(ts);
    assert(ts == "04:12:23");

    ts = sparen.formatInterval(t[1], "$+H:$M:$S.$F", 3);
    Log(ts);
    assert(ts == "04:12:23.340");

    ts = sparen.formatInterval(t[1], "$+H:$M:$S.$f", 3);
    Log(ts);
    assert(ts == "04:12:23.34");

    ts = sparen.formatInterval(t[2], "$+H:$M:$S");
    Log(ts);
    assert(ts == "2956:02:06");

    ts = sparen.formatInterval(t[2], "$d days, $H:$M:$S");
    Log(ts);
    assert(ts == "123 days, 04:02:06");

    ts = sparen.formatInterval(t[2], "$d days, $h hours, $m minutes, $s seconds");
    Log(ts);
    assert(ts == "123 days, 4 hours, 2 minutes, 6 seconds");

    ts = sparen.formatInterval(t[3], "$y years, $d days, $h hours, $m minutes, $s seconds");
    Log(ts);
    assert(ts == "4 years, 123 days, 4 hours, 2 minutes, 6 seconds");

    ts = sparen.formatInterval(25.001, "$s.$f");
    Log(ts);
    assert(ts == "25.001");

    ts = sparen.formatInterval(25.0015, "$s.$f");
    Log(ts);
    assert(ts == "25.002");

    ts = sparen.formatInterval(25.0012345, "$s.$f", 4);
    Log(ts);
    assert(ts == "25.0012");

    ts = sparen.formatInterval(25.0012, "$s.$F", 6);
    Log(ts);
    assert(ts == "25.001200");
}

function isCmd(lst, cmd)
{
    return !lst || !lst.length ||  0 <= `,${lst.toLowerCase()},`.indexOf(`,${cmd.toLowerCase()},`);
}

function main()
{
    let run = process.argv.slice(2).join(',');

    // Save to a disk file
    // sparen.obj.setLogFile('./sparen.log');

    // Output to stderr
    // sparen.obj.setLogFunction(console.error);

    Log(Fmt(sparen.__info__, null, 2));
    Log("--- START TESTS ---\n");

    // Run tests
    let tests = [test_1, test_2, test_3, test_4, test_5];
    for (let k in tests)
        if (isCmd(run, String(parseInt(k)+1)))
        {   Log('-----------------------------------------------------------');
            Log(` - ${tests[k].name}()`);
            Log('-----------------------------------------------------------\n');
            tests[k]();
        }

    Log('--- Done ---\n');
}

// Exit handling
process.on('exit',function() { Log('~ exit ~');});
process.on('SIGINT',function() { Log('~ keyboard ~'); process.exit(-1); });
process.on('uncaughtException',function(e) { Log('~ uncaught ~', e); process.exit(-1); });

// Run the program
main();



build.js
#!/usr/bin/env node
'use strict'

const fs = require('fs');
const path = require('path');
const Log = console.log;

function loadConfig(fname)
{   if (!fs.existsSync(fname))
        return {};
    let r = {};
    let data = fs.readFileSync(fname, 'utf8');
    let lines = data.replace(/\r\n/g, "\n").replace(/\r/g, "\n").split("\n");
    lines.forEach(v =>
        {   v = v.trim();
            if ('#' != v[0])
            {   let parts = v.split(/\s+/);
                if (1 < parts.length)
                {   let k = parts.shift().trim().toLowerCase();
                    r[k] = parts.join(' ');
                }
            }
        });
    return r;
}

function processFile(cfg, fin, fout, removeComments=true)
{   let data = fs.readFileSync(fin, 'utf8');
    if (removeComments)
        data = data.replace(/[^\n]*\/\/.*\n/g, '').replace(/[^\n]*#.*\n/g, '');
    const rx = /(%.*?%)/g
    for (let m; m = rx.exec(data); m)
        if (0 <= m.index)
        {   let k = m[0].replace(/%/g, '').toLowerCase();
            if (k in cfg)
                data = data.slice(0, m.index) + cfg[k] + data.slice(m.index+m[0].length);
        }
    fs.writeFileSync(fout, data);
}

function parseParams(args, merge={})
{
    let r = {'*':[]}, ref = null;
    for (let k in args)
    {   let v = args[k];
        if (v[0] == '-')
        {   if (v[1] != '-')
                for (let i = 1; i < v.length; i++)
                    ref = v[i], r[ref] = true;
            else
            {   let eq = v.indexOf('=');
                if (0 > eq)
                    ref = v.slice(2), r[ref] = true;
                else
                    r[v.slice(2, eq)] = v.slice(eq + 1), ref = null;
            }
        }
        else if (ref)
            r[ref] = v, ref = null;
        else
            r['*'].push(v);
    }
    for (let k in merge)
    {   let v = merge[k];
        if ((k in r) && !(v in r))
            r[v] = r[k];
        else if (!(k in r) && (v in r))
            r[k] = r[v];
    }
    return r;
}

function fileCopy(src, dst, opts={overwrite:true, recursive:true, show:Log})
{
    // Ensure the source exists
    if (!fs.existsSync(src))
    {   if (opts.show)
            opts.show(`[ignore] Not found: ${src}`);
        return;
    }

    // If the source is a file just copy it
    if (!fs.lstatSync(src).isDirectory())
    {   fs.copyFileSync(src, dst);
        if (opts.show)
            opts.show(`[copied] ${src} -> ${dst}`);
        return;
    }

    // Delete the destination if it exists
    if (fs.existsSync(dst))
    {   if (fs.lstatSync(dst).isDirectory())
            fs.rmdirSync(dst, { recursive: true });
        else
            fs.unlinkSync(dst);
    }

    // Create the destination directory
    if (!fs.existsSync(dst))
        fs.mkdirSync(dst);

    // Copy each file / directory
    let dirlist = fs.readdirSync(src);
    for (let i = 0; i < dirlist.length; i++)
    {
        let name = dirlist[i];
        let fsrc = path.join(src, name);
        let fdst = path.join(dst, name);
        let status = '';
        if (!opts.overwrite && fs.existsSync(fdst))
            status = 'exists';
        else
        {
            let stat;
            try { stat = fs.lstatSync(fsrc); }
            catch(e) { if (opts.show) opts.show(e); continue; }
            status = 'copied';
            if (!stat.isDirectory())
                fs.copyFileSync(fsrc, fdst);
            else if (opts.recursive)
                rCopy(fsrc, fdst);
            else
                status = 'ignore';
        }
        if (status && opts.show)
            opts.show(`[${status}] ${fsrc} -> ${fdst}`);
    }
}

function createDox()
{
    return new Promise((resolve, reject)=>
    {
        let doxygen = null;
        try { doxygen = require('doxygen'); }
        catch(e){ return reject(e); }

        Log("\Preparing doxygen...");

        let dcfg = path.join(dst, 'doxygen.cfg');
        return doxygen.downloadVersion()
            .then((data) =>
            {
                Log("\nGenerating Documentation...");
                var doxopts = {
                    PROJECT_NAME        : cfg.name,
                    OUTPUT_DIRECTORY    : path.join(src, 'dox'),
                    INPUT               : dst,
                    RECURSIVE           : "YES",
                    FILE_PATTERNS       : ["*.js"],
                    EXTENSION_MAPPING   : "js=Javascript",
                    GENERATE_LATEX      : "NO",
                    EXCLUDE_PATTERNS    : ["*/node_modules/*"]
                };
                doxygen.createConfig(doxopts, dcfg);
                return resolve(doxygen.run(dcfg));
            });
    });
}

function main()
{
    let _p = parseParams(process.argv, {'i':'install', 'g': 'global'});
    Log("Command", _p);

    let src = __dirname;
    let dst = path.join(src, 'dist');

    // Delete output directory if it already exists
    if (fs.existsSync(dst))
        fs.rmdirSync(dst, { recursive: true });

    // Create new output directory
    fs.mkdirSync(dst);
    if (!fs.existsSync(dst))
        throw `Failed to create output directory : ${dst}`;

    // Set working directory
    process.chdir(dst);

    // Load the config file
    let fcfg = path.join(src, 'PROJECT.txt');
    let cfg = loadConfig(fcfg);
    if (0 >= Object.keys(cfg).length)
        throw `Failed to load config: ${fcfg}`;

    // Create package file
    let pck_in = path.join(src, 'var', 'in.package.json');
    let pck_out = path.join(dst, 'package.json');
    processFile(cfg, pck_in, pck_out);
    if (!fs.existsSync(pck_out))
        throw `Failed to create package file : ${pck_out}`;

    // Copy install files
    for (let v of ['PROJECT.txt', 'README.md', 'LICENSE', 'lib', 'bin', 'test'])
        fileCopy(path.join(src, v), path.join(dst, v));

    // Installing?
    Log("\nPackaging...");

    const cp = require('child_process');

    let cmd = `npm pack`;
    cp.exec(cmd, { cwd: dst }, (error, stdout, stderr) =>
        {
            Log(stdout);
            Log(stderr);

            // Did we get a package
            let npkg = `${cfg.name}-${cfg.version}.tgz`
            let spkg = path.join(dst, npkg);
            if (!fs.existsSync(spkg))
                throw `Failed to create package : ${spkg}`;

            // Copy package file
            let rpkg = path.join(src, 'pkg');
            if (!fs.existsSync(rpkg))
                fs.mkdirSync(rpkg);

            // Copy package file
            let dpkg = path.join(rpkg, npkg);
            fs.renameSync(spkg, dpkg);
            fs.copyFileSync(dpkg, path.join(rpkg, `${cfg.name}-latest.tgz`));

            Log(`Created : ./pkg/${cfg.name}-latest.tgz -> ./pkg/${npkg}\n`);

            if (_p.install)
            {
                Log("\nInstalling...");
                let sudo = fs.existsSync('/usr/bin/sudo') ? 'sudo ' : '';
                let cmd = _p.g ? (sudo+`npm install -g ${dpkg}`) : `npm install ${dpkg}`;
                cp.exec(cmd, { cwd: dst }, (error, stdout, stderr) =>
                    {
                        Log(stdout);
                        Log(stderr);
                    });
            }
        });
}

// Exit handling
process.on('exit', function() {});
process.on('SIGINT', function() { Log('~ keyboard ~'); process.exit(-1); });
process.on('uncaughtException', function(e) { Log('~ uncaught ~', e); process.exit(-1); });

// Run the program
main();




lib/index.js
#!/usr/bin/env nodejs
'use strict';

const fs = require('fs');
const path = require('path');

function loadConfig(fname)
{   if (!fs.existsSync(fname))
        return {};
    let r = {};
    let data = fs.readFileSync(fname, 'utf8');
    let lines = data.replace(/\r\n/g, "\n").replace(/\r/g, "\n").split("\n");
    lines.forEach(v =>
        {   v = v.trim();
            if ('#' != v[0])
            {   let parts = v.split(/\s+/);
                if (1 < parts.length)
                {   let k = parts.shift().trim().toLowerCase();
                    r[k] = parts.join(' ');
                }
            }
        });
    return r;
}

const sparen = require('./sparen.js');
const canvas = require('./canvas.js');

var _g_sparen_obj = new sparen.Sparen();
module.exports =
{
    __info__    : loadConfig(path.join(path.dirname(__dirname), 'PROJECT.txt')),
    obj         : _g_sparen_obj,
    log         : _g_sparen_obj.log,

    ...sparen,
    ...canvas
}


lib/sparen.js
#!/usr/bin/env nodejs
'use strict';

const fs = require('fs');
const path = require('path');

/** Exports
*/
module.exports = {
        Sparen          : Sparen,
        initArray       : initArray,
        plotArray       : plotArray,
        formatInterval  : formatInterval
    };

/** Initialize an array with return value of provided function
 * @param [in] sz   - Array size
 * @param [in] f    - Function or value that initializes each array element
 *
 * @return An array of size sz, with each element initialized with f().
*/
function initArray(sz, f)
{
    // Try ES6
    // try { return Array.from({length: sz}, f); } catch(e) {}

    let a = Array(sz);
    if (typeof f === 'function')
        for (let i = 0; i < sz; i++)
            a[i] = f(i);
    else
        for (let i = 0; i < sz; i++)
            a[i] = f;
    return a;
}


/** Logging class
    @param [in] linetmpl    - Line templates

                                ts          = timestamp
                                file        = filename
                                line        = line number
                                function    = Function name

                                Examples:
                                    '[<<ts>>] <<file>>(<<line>>): '                 -> [10:06:37] ./test.py(147):
                                    '[<<ts>>] <<file>>::<<function>>(<<line>>): '   -> [10:08:47] ./test.py::test_5(147):

    @param [in] tstmpl      - Timestamp template
                                Examples:
                                    '%H:%M:%S'

    @param [in] reldir      - Relative directory to files to show in logs
                                null    = Full paths
                                '-'     = No paths
                                '.'     = Relative to current working directory
                                ...     = Relative to specified path

*/
function Sparen(linetmpl=null, tstmpl='%H:%M:%S', reldir='.')
{
    //--------------------------------------------------------------
    // Interface

    this.log = log;
    this.addLogFilter = addLogFilter;
    this.addLogFilters = addLogFilters;
    this.getLogFilters = getLogFilters;
    this.setLogFunction = setLogFunction;
    this.setLogFile = setLogFile;
    this.setMinLogInterval = setMinLogInterval;


    //--------------------------------------------------------------
    // Data

    let m_color_filters = {};
    let m_colors = {
        'BLACK'     : '\u001b[90m',
        'RED'       : '\u001b[91m',
        'GREEN'     : '\u001b[92m',
        'YELLOW'    : '\u001b[93m',
        'BLUE'      : '\u001b[94m',
        'MAGENTA'   : '\u001b[95m',
        'CYAN'      : '\u001b[96m',
        'WHITE'     : '\u001b[97m',
        'BOLD'      : '\u001b[1m',
        'FAINT'     : '\u001b[2m',
        'ITALIC'    : '\u001b[3m',
        'UNDERLINE' : '\u001b[4m',
        'BLINK'     : '\u001b[5m',
        'NEGATIVE'  : '\u001b[7m',
        'STRIKEOUT' : '\u001b[9m',
        'DEFAULT'   : '\u001b[0m'
    };
    let m_log = console.log;
    let m_logFile = null;
    let m_logFileFailTime = 0;
    let m_minLogInterval = 60;
    let m_linetmpl = linetmpl;
    let m_reldir = null;

    if ('.' == reldir)
        m_reldir = process.cwd();
    else
        m_reldir = reldir;


    //--------------------------------------------------------------
    // Functions

    /// Returns the current stack
    function getStack()
    {
        let ret = null;
        let orig = Error.prepareStackTrace;
        try
        {
            Error.prepareStackTrace = (_, stack) => {
                return stack;
            };
            let err = new Error;
            Error.captureStackTrace(err, log);
            ret = err.stack;
        }
        catch(e) {}
        Error.prepareStackTrace = orig;
        return ret;
    }

    /// Parse filename from string
    function parseFileName(p)
    {   let _p = String(p).split("/");
        return _p.length ? _p[_p.length - 1] : p;
    }

    /// Set log output function / Default is console.log
    function setLogFunction(fLog)
    {   m_log = fLog;
    }

    /// Set a log filename for log output
    function setLogFile(fname)
    {   m_logFile = fname;
    }

    /// Set the minimum time between logging internal errors
    function setMinLogInterval(minLogInterval)
    {   m_minLogInterval = minLogInterval;
    }

    /**
     * @param [in] *    - All parameters are logged
    */
    function log()
    {
        // Stack information
        let stack = getStack();
        let st = stack ? stack[0] : null;
        let fn = st ? st.getFunctionName() : '';
        let fpath = st ? st.getFileName() : '';
        let fname = st ? parseFileName(st.getFileName()) : '';
        let lin = st ? st.getLineNumber() : '';
        let file = fname;
        if (m_reldir)
            if ('-' == m_reldir)
                file = fname;
            else if ('+' == m_reldir)
                file = fpath;
            else
            {   let usefull = false;
                try { if (fpath.split('/')[1] != m_reldir.split('/')[1]) usefull = true;
                } catch(e) {usefull = false;}
                if (usefull)
                    file = fpath;
                else
                    file = "./" + path.relative(m_reldir, fpath);
            }
        else
            file = fpath;

        // Format arguments
        let args = [];
        for (let k in arguments)
        {
            let v = arguments[k];
            try
            {
                if (v && typeof(v) == 'object')
                {   if (v.stack)
                        args.push("\n " + v.stack + "\n");
                    if (0 < Object.keys(v).length)
                        args.push(JSON.stringify(v));
                }
                else
                    args.push(String(v));
            }
            catch(e)
            {
                args.push(String(v));
            }
        }

        args = args.join(' ');

        // Do we have a true console
        let isTty = true;
        try { isTty = process.stdout.isTTY ? true : false; }
        catch(e)
        {   try
            {   const tty = require('tty');
                if (!tty.isatty(process.stdout.fd))
                    isTty = false;
            } catch(e) {}
        }

        // Apply color filters
        let beg = '', end = '';
        try
        {
            if (isTty)
                for (let k in m_color_filters)
                {   let v = m_color_filters[k];
                    if (0 <= args.indexOf(k))
                    {   end = m_colors['DEFAULT'];
                        for (let _c in v)
                        {   let c = v[_c];
                            c = c.toUpperCase();
                            if ('BLOCK' == c)
                                return;
                            if (c in m_colors)
                                beg += m_colors[c];
                        }
                    }
                }
        }
        catch(e)
        {
            m_log(e);
            beg = '';
            end = '';
        }

        // Create timestamp
        let now = new Date();
        let hour = String(now.getHours()).padStart(2,'0');
        let mins = String(now.getMinutes()).padStart(2,'0');
        let secs = String(now.getSeconds()).padStart(2,'0');
        let ts = `${hour}:${mins}:${secs}`;
        let tstr = isTty ? `\u001b[2m[${ts}]\u001b[0m` : `[${ts}]`

        // Message string
        let lstr;
        if (m_linetmpl)
            lstr = m_linetmpl.replace('<<ts>>', ts)
                             .replace('<<file>>', file)
                             .replace('<<function>>', fn)
                             .replace('<<line>>', lin);
        else
            lstr = `[${ts}] ${file}(${lin}): `;

        // Show the message
        m_log(`\u001b[2m${lstr}\u001b[0m${beg}${args}${end}`);

        // // Message string
        lstr += `${args}\n`;

        // Save to file if a name was provided
        if (m_logFile)
        {
            function logFileError(e)
            {   let t = Date.now() / 1000;
                if (t < m_logFileFailTime)
                    return;
                m_logFileFailTime = t + m_minLogInterval;
                m_log(e);
            }

            // try { fs.appendFile(m_logFile, lstr, logFileError); }
            try { fs.appendFileSync(m_logFile, lstr); }
            catch(e) { logFileError(e); }
        }

        return lstr;
    }

    /** Adds a single filter to colorize text
        @param [in] f   - Case insensitive filter string,
                          values can be any of
                            black, red, green, yellow, blue,
                            magenta, cyan, white, bold, faint,
                            italic, underline, blink, strikeout

        Example:
        @begincode

            // Color all strings containing 'error' red
            addLogFilter("error:red")

        @endcode
    */
    function addLogFilter(f)
    {
        let p = f.split(':');
        if (1 < p.length)
        {   try
            {   ''.search(p[0]);
                m_color_filters[p[0]] = p.slice(1);
            }
            catch(e) { m_log("ERROR: Invalid log filter:", String(e)); }
        }
        else if (f in m_color_filters)
            delete m_color_filters[f];
    }

    /** Adds a list of filters to colorize text
        @param [in] f   - Comma separated case insensitive filter strings
                          values can be any of
                            black, red, green, yellow, blue,
                            magenta, cyan, white, bold, faint,
                            italic, underline, blink, strikeout

        Example:
        @begincode

            // Color all strings containing 'error' red, 'warning' yellow,
            // Texts containing 'info' will blink
            addLogFilter("error:red,warning:yellow,info:blink")

        @endcode
    */
     function addLogFilters(f)
    {
        let p = f.split(',');
        for(let v in p)
            if (p[v])
                addLogFilter(p[v]);
    }

    /// Returns current log filters
    function getLogFilters()
    {
        return m_color_filters;
    }
}

/** Plots an array
    @param [in] a       - Array data to plot
    @param [in] fn      - Optional function for retrieving data elements
    @param [in] scalex  - True to scale values to the x axis
    @param [in] scaley  - True to scale values to the y axis
    @param [in] height  - Height of the output plot
    @param [in] width   - Width of the output plot
    @param [in] plot    - Character to use to plot the data
    @param [in] miny    - Minimum Y axis value
    @param [in] maxy    - Maximum Y axis value
    @param [in] marginy - Y axis top and bottom margin
*/
function plotArray(a, fn = null, scalex = true, scaley = true, height = 12, width = 70,
                   plot='.', miny = 0, maxy = 100, marginy = 1)
{
    let w = width;
    let h = height;
    let m = initArray(h, ()=>initArray(w, " "));

    let l = a.length;
    let v = fn ? a.map(fn) : a;

    // Scaling the y axis?
    if (scaley)
    {   miny = Math.min.apply(null,v) - marginy;
        maxy = Math.max.apply(null,v) + marginy;
    }
    let rg = maxy - miny;
    if (0 == rg)
        rg = 1;

    let x = 0, xa = 0;
    for (let i = 0; i < l; i++)
    {
        // New y value
        let y = parseInt((v[i]-miny) * h / rg);

        // Update x pos
        while (xa >= l)
            xa -= l, x += 1;
        if (x >= w)
            break;

        // Incremental division
        xa += scalex ? w : l;

        // Character to plot
        let pt = plot;
        if (y < 0)
            y = 0, pt = '_';
        if (y >= h)
            y = h - 1, pt = '^';

        // Plot the character
        m[h - y - 1][x] = pt;
    }

    // How wide is the y gutter
    let j = Math.max(String(miny).length, String(maxy).length);

    // Build the plot
    let y = h - 1;
    let pl = null, ret = [];
    for (let r of m)
    {   let py = miny + ((y+1) * rg / h);
        py = parseInt(py - (0 > py ? 1 : 0));
        py = String(py).padStart(j, ' ');
        let sy = pl == py ? '  ' : py;
        pl = py;
        ret.push(`${sy} : ${r.join('')}`);
        y -= 1
    }

    // Build the x axis markers
    ret.push(initArray(j+1, ' ').join('') + initArray(w+2, '-').join(''));
    let nstr = '';
    for (let x = 0; x < parseInt(w/10); x++)
        nstr += String((x+1)*10).padStart(9, ' ')+'^';
    ret.push(initArray(j, ' ').join('') + nstr);

    return ret.join('\n');
}

/** Returns a formated string of the specified time interval
    @param [in] t   - Time interval in seconds
    @param [in] fmt - Format string
                        $s  - Seconds
                        $S  - Seconds with leading zero
                        $m  - Minutes
                        $M  - Minutes with leading zero
                        $h  - Hours
                        $H  - Hours with leading zero
                        $d  - Days
                        $D  - Days with leading zero
                        $y  - Years
                        $Y  - Years with leading zero
                        $f  - Decimal
                        $F  - Decimal with trailing zeros
                        $+_ - Show full value

    @param [in] dec - Number of decimal digits to include

    @begincode

        print(formatInterval())

    @encode
*/
function formatInterval(t, fmt="$+H:$M:$S", dec=3)
{
    let rep = {
        "$f":   (t) => (t-parseInt(t)).toFixed(dec).slice(2).replace(/0+$/g, ''),
        "$F":   (t) => (t-parseInt(t)).toFixed(dec).slice(2),
        "$s":   (t) => String(parseInt(t) % 60),
        "$+s":  (t) => String(parseInt(t)),
        "$S":   (t) => String(parseInt(t) % 60).padStart(2,'0'),
        "$+S":  (t) => String(parseInt(t)).padStart(2,'0'),
        "$m":   (t) => String(parseInt(t / 60) % 60),
        "$+m":  (t) => String(parseInt(t / 60)),
        "$M":   (t) => String(parseInt(t / 60) % 60).padStart(2,'0'),
        "$+M":  (t) => String(parseInt(t / 60)).padStart(2,'0'),
        "$h":   (t) => String(parseInt(t / 60 / 60) % 24),
        "$+h":  (t) => String(parseInt(t / 60 / 60)),
        "$H":   (t) => String(parseInt(t / 60 / 60) % 24).padStart(2,'0'),
        "$+H":  (t) => String(parseInt(t / 60 / 60)).padStart(2,'0'),
        "$d":   (t) => String(parseInt(t / 24 / 60 / 60) % 365),
        "$+d":  (t) => String(parseInt(t / 24 / 60 / 60)),
        "$D":   (t) => String(parseInt(t / 24 / 60 / 60) % 365).padStart(3,'0'),
        "$+D":  (t) => String(parseInt(t / 24 / 60 / 60)).padStart(3,'0'),
        "$y":   (t) => String(parseInt(t / 365 / 24 / 60 / 60)),
        "$Y":   (t) => String(parseInt(t / 365 / 24 / 60 / 60)).padStart(2,'0'),
    }

    let max = 99;
    for (let k in rep)
        while (0 < max-- && 0 <= fmt.indexOf(k))
            fmt = fmt.replace(k, rep[k](t));

    return fmt;
}
